File: Python/Binary Tree Maximum Path Sum.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return an integer
    value = float('-inf')
    def maxPathSum(self, root):
        self.helper(root)
        return self.value
        
    def helper(self, root ):
        if not root: return 0
        left = max(0, self.helper(root.left))
        right = max(0, self.helper(root.right))
        self.value = max(self.value, left+right+root.val)
        return max(left, right) + root.val
        

File: Java/src/Binary Tree Maximum Path Sum.java
import javax.swing.tree.TreeNode;

public class Solution
{
  int value;

  public int maxPathSum(TreeNode root)
  {
    value = Integer.MIN_VALUE;
    helper(root);
    return value;
  }

  private int helper(TreeNode node)
  {
    if (node == null)
      return 0;
    int left = Math.max(0, helper(node.left));
    int right = Math.max(0, helper(node.right));
    value = Math.max(value, left + right + node.val);
    return Math.max(left, right) + node.val;
  }
}
